handle errors in ProcessGroupAgent::listenLoop().

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32957 handle errors in ProcessGroupAgent::listenLoop().**

Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

rohan-varma committed:
[wip][rpc] handle errors in ProcessGroupAgent::listenLoop().

Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown.

This PR also adds a new option, `processGroupTimeout` to PG agent's backend
options. This allows us to control the gloo timeout.

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[wip][rpc] handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown.

This PR also adds a new option, `processGroupTimeout` to PG agent's backend
options. This allows us to control the gloo timeout.

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[wip][rpc] handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown.

This PR also adds a new option, `processGroupTimeout` to PG agent's backend
options. This allows us to control the gloo timeout.

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

rohan-varma committed:
Update on "handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown.

This PR also adds a new option, `processGroupTimeout` to PG agent's backend
options. This allows us to control the gloo timeout.

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit a8fdbf01:


**None of the build failures appear to be your fault**.

* **1/1** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=a8fdbf0110d33a538cf29fd16c66c2c1d858d2b9).

### :snowflake: 1 failure recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4407633) pytorch_macos_10_13_py3_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=94843698) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=94843698)) :snowflake:
<details>
<summary>
<code>Feb 03 18:09:36 RuntimeError: Process 0 terminated or timed out after 100.05249500274658 seconds</code>
</summary>

```
Feb 03 18:09:36 ====================================================================== 
Feb 03 18:09:36 ERROR [100.075s]: test_listenloop_error (__main__.RpcTestWithSpawn) 
Feb 03 18:09:36 ---------------------------------------------------------------------- 
Feb 03 18:09:36 Traceback (most recent call last): 
Feb 03 18:09:36   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 174, in wrapper 
Feb 03 18:09:36     self._join_processes(fn) 
Feb 03 18:09:36   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 255, in _join_processes 
Feb 03 18:09:36     self._check_return_codes(elapsed_time) 
Feb 03 18:09:36   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 274, in _check_return_codes 
Feb 03 18:09:36     raise RuntimeError('Process {} terminated or timed out after {} seconds'.format(i, elapsed_time)) 
Feb 03 18:09:36 RuntimeError: Process 0 terminated or timed out after 100.05249500274658 seconds 
Feb 03 18:09:36  
Feb 03 18:09:36 ---------------------------------------------------------------------- 
Feb 03 18:09:36 Ran 87 tests in 197.278s 
Feb 03 18:09:36  
Feb 03 18:09:36 FAILED (errors=1, skipped=2) 
Feb 03 18:09:36  
Feb 03 18:09:36 Generating XML reports... 
Feb 03 18:09:36 Traceback (most recent call last): 
Feb 03 18:09:36   File "test/run_test.py", line 457, in <module> 
Feb 03 18:09:36     main() 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -542,7 +546,22 @@ void ProcessGroupAgent::listenLoop() {
+    bool aborted;
+    try {
+      aborted = !work->wait();
+    } catch (const gloo::IoException& e) {
```
As I mentioned in https://github.com/pytorch/pytorch/pull/32602 can we use the `listenLoop` and `listenLoopInternal` mechanism to catch all exceptions? For example, `recvAnysource` on line 543 can throw exceptions as well, but we don't catch them in this PR.

As you mentioned in https://github.com/pytorch/pytorch/pull/32602, setting rpcRunning_ to false is not a good idea, but let's set some error flag in ProcessGroupAgent and call ProcessGroupAgent::shutdown to shutdown the RPC framework. Then in ProcessGroupAgent.send() we can throw an appropriate exception if listenLoop crashed due to an exception.

rohan-varma replied:
Not sure if it will work to call `::shutdown()` from the listening thread. `::shutdown()` joins the listening thread itself, which could lead to issues since we indirectly have a thread call join on itself.

pritamdamania87 commented on torch/testing/_internal/dist_utils.py:
```diff
@@ -87,17 +87,25 @@ def new_test_method(self, *arg, **kwargs):
     '''
     while True:
         try:
+            time.sleep(sleep_duration)
```
nit: Usually better to sleep after the operation rather than sleeping the first time we enter the loop.

rohan-varma replied:
I changed this to sleep first otherwise for the UT I added, we'll send an RPC to node 0 which we're waiting for to timeout. As a result the timer in the ProcessGroup will be reset and we'll be forced to do a second sleep, so this approach allows us to save ~3s on this unit test at a minor cost (0.1 sleep) to other unit tests that use this function.

pritamdamania87 commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1711,3 +1711,49 @@ def test_function_not_on_callee(self):
+        backend_opts = self.rpc_backend_options
+        # Set a small process group timeout on node 1.
+        rank_1_pg_timeout = timedelta(seconds=1)
+        pg_sleep_interval = rank_1_pg_timeout.seconds * 3
```
nit: Define this on line 1741 (closer to where its used).

pritamdamania87 commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1711,3 +1711,49 @@ def test_function_not_on_callee(self):
+                with self.assertRaisesRegex(
+                    RuntimeError, "Application timeout caused pair closure"
+                ):
+                    rpc.rpc_async("worker0", torch.add, args=(1, 1)).wait()
```
nit: Any reason we can't just use rpc_sync?

pritamdamania87 commented on torch/distributed/rpc/constants.py:
```diff
@@ -8,3 +8,5 @@
 # For ProcessGroupAgent.
 DEFAULT_NUM_SEND_RECV_THREADS = 4
+# Same default timeout as in c10d.
+DEFAULT_PROCESS_GROUP_TIMEOUT = _default_pg_timeout
```
Does this mean `pg->recvAnySource` will throw an exception if no data is received on the wire for 30 minutes? It seems like we should default this to infinity to ensure `recvAnySource` never throws a timed out exception.

If what I mentioned above recvAnySource is true, then we shouldn't allow users to specify `process_group_timeout`, since they can just set it to a low value and this would cause listenLoop to terminate. For the unit test, we could have a private API `_set_process_group_timeout_for_testing` which sets an appropriate value for an internal flag that we check before calling `init_process_group` (process_group_timeout shouldn't be part of RpcBackendOptions).

rohan-varma replied:
Yes, this is true. It can be seen by setting a short process_group_timeout such as 5s, allowing RPC init to complete, and then just waiting. Currently listenLoop will crash since there was no data sent over the wire. Should we bump this to be infinite?

rohan-varma replied:
Filed https://github.com/pytorch/pytorch/issues/33107 to track this

rohan-varma committed:
Update on "handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown.

This PR also adds a new option, `processGroupTimeout` to PG agent's backend
options. This allows us to control the gloo timeout.

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

rohan-varma committed:
Update on "handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown.

This PR also adds a new option, `processGroupTimeout` to PG agent's backend
options. This allows us to control the gloo timeout.

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

rohan-varma committed:
Update on "handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown.

This PR also adds a new option, `processGroupTimeout` to PG agent's backend
options. This allows us to control the gloo timeout.

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

rohan-varma committed:
Update on "handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown.

This PR also adds a new option, `processGroupTimeout` to PG agent's backend
options. This allows us to control the gloo timeout.

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

rohan-varma committed:
Update on "handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown.

This PR also adds a new option, `processGroupTimeout` to PG agent's backend
options. This allows us to control the gloo timeout.

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

rohan-varma committed:
Update on "handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown.

This PR also adds a new option, `processGroupTimeout` to PG agent's backend
options. This allows us to control the gloo timeout.

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

rohan-varma committed:
Update on "handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown.

This PR also adds a new option, `processGroupTimeout` to PG agent's backend
options. This allows us to control the gloo timeout.

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

rohan-varma committed:
Update on "handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown.

This PR also adds a new option, `processGroupTimeout` to PG agent's backend
options. This allows us to control the gloo timeout.

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

rohan-varma committed:
Update on "handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown.

This PR also adds a new option, `processGroupTimeout` to PG agent's backend
options. This allows us to control the gloo timeout.

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

rohan-varma committed:
Update on "handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown.

This PR also adds a new option, `processGroupTimeout` to PG agent's backend
options. This allows us to control the gloo timeout.

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

rohan-varma committed:
Update on "handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown.

This PR also adds a new option, `processGroupTimeout` to PG agent's backend
options. This allows us to control the gloo timeout.

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

rohan-varma commented:
@pritamdamania87 Addressed the CR comments, though I don't think it's a good idea to commit the UT since it can be flaky. I investigate the idea of setting the timeout for individual `recvs()` after initialization, but this requires adding some hacky hooks, and then waiting until the thread in ProcessGroupAgent has called `wait()` with the proper argument being used, which can still be flaky.

To unblock for now I've added a C++ unit test in the internal PG agent tests (see internal diff), which isn't ideal since those are not currently OSS. We should work on exporting them to OSS.

rohan-varma committed:
Update on "handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown.

This PR also adds a new option, `processGroupTimeout` to PG agent's backend
options. This allows us to control the gloo timeout.

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

rohan-varma committed:
Update on "handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown.

This PR also adds a new option, `processGroupTimeout` to PG agent's backend
options. This allows us to control the gloo timeout.

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

rohan-varma committed:
Update on "handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown.

This PR also adds a new option, `processGroupTimeout` to PG agent's backend
options. This allows us to control the gloo timeout.

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

rohan-varma committed:
Update on "handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown.

This PR also adds a new option, `processGroupTimeout` to PG agent's backend
options. This allows us to control the gloo timeout.

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

rohan-varma committed:
Update on "handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

rohan-varma committed:
Update on "handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

rohan-varma committed:
Update on "handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

rohan-varma committed:
Update on "handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

rohan-varma committed:
Update on "handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -267,6 +264,14 @@ void ProcessGroupAgent::shutdown() {
+  {
+    std::unique_lock<std::mutex> guard(listenLoopExceptionMutex_);
+    if (listenLoopException_) {
+      guard.unlock();
```
nit: Do we need this unlock? std::rethrow will automatically trigger the unlock

rohan-varma committed:
Update on "handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

rohan-varma committed:
Update on "handle errors in ProcessGroupAgent::listenLoop()."


Closes https://github.com/pytorch/pytorch/issues/29703. If there is a
gloo timeout and `recvWork->wait()` times out in `listenLoop()`,
processGroupagent crashes since there is an unhandled exception in a thread.
This catches the exception and exits the listen loop. In a follow up diff, we
will enhance these error conditions so that if users attempt to send RPCs
again, they are notified that the RPC agent was in a bad state and it was
shutdown

Differential Revision: [D19678979](https://our.internmc.facebook.com/intern/diff/D19678979/)

[ghstack-poisoned]

osalpekar commented on torch/csrc/distributed/rpc/process_group_agent.h:
```diff
@@ -138,8 +138,15 @@ class ProcessGroupAgent : public RpcAgent {
-  // receiving messages
+  // Loop for receiving messages. Calls listenLoopInternal and handles errors
+  // such as timeouts on the process group.
+  virtual void listenLoopInternal();
```
Does this function need to be virtual? Seems like the only child classes of ProcessGroupAgent are those in part of tests and don't override listen loop.

rohan-varma replied:
Right, initially I had some testing that overrides it but I removed that. This PR is already landing though so I'll fix it in a followup
